### PR TITLE
upgrade ray-tracing branch to main branch

### PR DESCRIPTION
**Pull request by Vecvec**
**Connections**
upgrades this branch to wgpu trunk, also fixes the tlas memory leak for me

**Description**
This branch is missing a lot of commits, this PR adds those, and also brings the tlas and blas structs up to work with those changes.

**Testing**
Tested with the three examples added by this change, as well as a basic program that (in a loop) creates blas and tlas, and immediately frees them.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown` (done by github actions)
  - [x] `--target wasm32-unknown-emscripten` (done by github actions)
- [x] Run `cargo xtask test` to run tests. (done by github actions)
- Add change to `CHANGELOG.md`. I do not think this is applicable as this is part of the ray-tracing PR (gfx-rs/wgpu#3631).

Now ready for review

### CODE REVIEW

**Comment by JMS55** [comment]
@Vecvec I would love to have this in wgpu. Given that @daniel-keitel seems to be busy atm, perhaps you want to open your own PR in the wgpu repo? (ofc crediting daniel-keitel).

**Comment by Vecvec** [comment]
Thank you! I had assumed @daniel-keitel was just away for Christmas and was not rushing. Currently I am away too and so cannot check merges with trunk have not broken anything, which I would like to do before opening a new PR. If @daniel-keitel is still away when I get back I will create a PR in wgpu as you suggest.

**Comment by JMS55** [comment]
No worries, thanks! Also, I'd love to see BLAS compaction support if you feel like implementing it. It's a major missing part of RT.

**Comment by Vecvec** [comment]
I'd love to implement this, however looking at Vulkan specs [VkCopyAccelerationStructureInfoKHR](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#VkCopyAccelerationStructureInfoKHR) it seems like this needs another hal function, because the suggested specification says :
> [Function compact_blas] Creates a new blas and copies (in a compacting way) the contents of the provided blas into the new one (compaction flag must be set). 

(https://github.com/gfx-rs/wgpu/issues/1040#issuecomment-1484274170)
Should something such as copy_acceleration_structure_to_acceleration_structure be implemented or is there some other way I have missed that does this?

**Comment by JMS55** [comment]
Yes, this will need some additional work in wgpu-hal. Should hopefully be pretty straightforward for that part, as wgpu-hal tends to be a simple wrapper over VK/DX/Metal. 
See https://developer.nvidia.com/blog/tips-acceleration-structure-compaction and https://nvpro-samples.github.io/vk_raytracing_tutorial_KHR/#accelerationstructure/bottom-levelaccelerationstructure

**Comment by Vecvec** [comment]
Something like this for Vulkan?
````rust
unsafe fn copy_acceleration_structure_to_acceleration_structure<T>(&mut self, src: &Api::AccelerationStructure, dst: &Api::AccelerationStructure, copy: AccelerationStructureCopy) {
        let ray_tracing_functions = self
            .device
            .extension_fns
            .ray_tracing
            .as_ref()
            .expect("Feature `RAY_TRACING` not enabled");

        let structure_type = match copy.type_flags {
            AccelerationStructureType::Triangles => vk::StructureType::ACCELERATION_STRUCTURE_GEOMETRY_TRIANGLES_DATA_KHR,
            AccelerationStructureType::AABBs => vk::StructureType::ACCELERATION_STRUCTURE_GEOMETRY_AABBS_DATA_KHR,
            AccelerationStructureType::Instances => vk::StructureType::ACCELERATION_STRUCTURE_GEOMETRY_INSTANCES_DATA_KHR,
        };

        let mode = match copy.copy_flags {
            AccelerationStructureCopy::Clone => vk::CopyAccelerationStructureModeKHR::CLONE,
            AccelerationStructureCopy::Compact => vk::CopyAccelerationStructureModeKHR::COMPACT,
        };

        unsafe {
            ray_tracing_functions.acceleration_structure.copy_acceleration_structure(
                vk::DeferredOperationKHR::null(),
                &vk::CopyAccelerationStructureInfoKHR {
                    s_type: structure_type,
                    p_next: std::ptr::null(),
                    src: src.raw,
                    dst: dst.raw,
                    mode,
                }
            ).expect("Copy Failed");
        }
    }
````

**Comment by JMS55** [comment]
Something like that, yeah. I think you also need to add support for a BLAS size query, in order for the user to be able to get what size the compacted BLAS should be.

**Comment by Vecvec** [comment]
And a size query like this?
````rust
unsafe fn get_acceleration_structure_compact_size(&self, acceleration_structure: &super::AccelerationStructure) -> wgt::BufferAddress {
        let ray_tracing_functions = self
            .shared
            .extension_fns
            .ray_tracing
            .as_ref()
            .expect("Feature `RAY_TRACING` not enabled");
        let mut size = vk::DeviceSize::default().to_ne_bytes();

        unsafe {
            ray_tracing_functions
                .acceleration_structure
                .write_acceleration_structures_properties(
                    &[acceleration_structure.raw],
                    vk::QueryType::ACCELERATION_STRUCTURE_COMPACTED_SIZE_KHR,
                    &mut size,
                    std::mem::size_of::<vk::DeviceSize>(),
                ).expect("getting properties failed");
        }
        wgt::BufferAddress::from_ne_bytes(size)
    }
````

**Comment by JMS55** [comment]
Probably. I've also just noticed there's a way to serialize and deserialize BLAS's, which is very cool. That would be nice to have too, although less necessary. You should probably talk to the wgpu devs for specific wpgu-related advice though, I'm not a maintainer, and haven't written much code for it.

**Comment by daniel-keitel** [comment]
Hi, I will be fully back from my Christmas break tomorrow (with access to my desktop pc).
I will look over your changes this week.
After that we should, coordinate what we want to implement. (@Vecvec  I would preferer talking over Discord, if it is ok for you)
As far as I can tell a drastic change in the Api will be necessary to be idiomatic in reference to Webgpu.

**Comment by Vecvec** [comment]
That's great! Hope you had a good holiday. 
I'm sorry but I don't have discord (and no intention of creating an account).
How drastic a change are you contemplating?

**Comment by Vecvec** [comment]
I have also put the compact blas function onto a separate branch (vecvec:ray-tracing-compaction), as I don't think it's in the scope of this PR.

**Comment by daniel-keitel** [comment]
@Vecvec the main change I am contemplating, is to somehow make the information stored in the creation info part of the wgpu-core library and reference count all required resources.

The problem at the moment is, that a reference to Blas is required to build/update the blas, so lifetimes get in the way sometimes.
Another problem is that acceleration structures don't keep there resources alive. (You get an error if you drop a vertex buffer used in a ac, when it is used).
I think it would be better (and more performant to validate), if it worked like Texture and TextureView, where the view keeps the texture alive if the Texture is dropped.

File: examples/src/ray_scene/mod.rs [hunk 1]
```diff
@@ -298,9 +298,8 @@ fn upload_scene_components(
 
 fn load_scene(device: &wgpu::Device, queue: &wgpu::Queue) -> SceneComponents {
     let mut scene = RawSceneComponents::default();
-    load_model(&mut scene, "/examples/skybox/models/teslacyberv3.0.obj");
 
-    load_model(&mut scene, "/examples/ray-scene/cube.obj");
+    load_model(&mut scene, "/ray_scene/cube.obj");
```
Comment (daniel-keitel) [inline examples/src/ray_scene/mod.rs:302]:
```suggestion
    load_model(&mut scene, "/skybox/models/teslacyberv3.0.obj");
```

File: examples/src/ray_scene/mod.rs [hunk 2]
```diff
@@ -482,10 +475,14 @@ impl wgpu_example::framework::Example for Example {
                 for y in 0..side_count {
                     let instance = self
                         .tlas_package
-                        .get_mut_single((x + y * side_count) as usize)
+                        .get_mut_single((x + y) * side_count)
                         .unwrap();
```
Comment (daniel-keitel) [inline examples/src/ray_scene/mod.rs:479]: With this change the example should behave like before
```suggestion
                        .get_mut_single(x + y * side_count)
```

**Pull request merged**